src/scaled-model.jl

```diff
@@ -0,0 +1,236 @@
+
+function scale_model!(scaling::IpoptScaling{T}, nlp) where T
+  n, m = NLPModels.get_nvar(nlp), NLPModels.get_ncon(nlp)
+  nnzj = NLPModels.get_nnzj(nlp)
+  x0 = NLPModels.get_x0(nlp)
+  g = NLPModels.grad(nlp, x0)
+  scaling_obj = min(one(T), scaling.max_gradient / norm(g, Inf))
+  scaling_cons = similar(x0, m)
+  scaling_jac  = similar(x0, nnzj)
+  fill!(scaling_cons, zero(T))
+  Ji, Jj = NLPModels.jac_structure(nlp)
+  NLPModels.jac_coord!(nlp, x0, scaling_jac)
+  _set_constraints_scaling!(scaling_cons, Ji, Jj, scaling_jac, scaling.max_gradient)
+  _set_jacobian_scaling!(scaling_jac, Ji, Jj, scaling_cons)
+  return (scaling_obj, scaling_cons, scaling_jac)
+end
+
+@doc raw"""
+    ScaledModel
+
+Scale the nonlinear program
+```math
+\begin{aligned}
+       min_x  \quad & f(x)\\
+\mathrm{s.t.} \quad &  c♭ ≤ c(x) ≤ c♯ \\
+                    & x ≥ 0
+\end{aligned}
+```
+as
+```math
+\begin{aligned}
+       min_x  \quad & σf . f(x)\\
+\mathrm{s.t.} \quad &  σc . c♭ ≤ σc . c(x) ≤ σc . c♯ \\
+                    & x ≥ 0
+\end{aligned}
+```
+with ``σf`` a scalar defined as
+```
+σf = min(1, max_gradient / norm(g0, Inf))
+
+```
+and ``σc`` a vector whose size is equal to the number of constraints in the model.
+For ``i=1, ..., m``,
+```
+σc[i] = min(1, max_gradient / norm(J0[i, :], Inf))
+
+```
+
+The vector ``g0 = ∇f(x0)`` and the matrix ``J0 = ∇c(x0)`` are resp.
+the gradient and the Jacobian evaluated at the initial point ``x0``.
+
+"""
+struct ScaledModel{T, S, M} <: NLPModels.AbstractNLPModel{T, S}
+  nlp::M
+  meta::NLPModels.NLPModelMeta{T, S}
+  counters::NLPModels.Counters
+  scaling_obj::T
+  scaling_cons::S # [size m]
+  scaling_jac::S  # [size nnzj]
+  buffer_cons::S  # [size m]
+end
+
+NLPModels.show_header(io::IO, nlp::ScaledModel) =
+  println(io, "ScaledModel - Model with scaled objective and constraints")
+
+
+function ScaledModel(
+    nlp::NLPModels.AbstractNLPModel{T, S};
+    scaling=IpoptScaling(T(100)),
+) where {T, S}
+  n, m = NLPModels.get_nvar(nlp), NLPModels.get_ncon(nlp)
+  x0 = NLPModels.get_x0(nlp)
+  buffer_cons  = S(undef, m)
+  scaling_obj, scaling_cons, scaling_jac = scale_model!(scaling, nlp)
+  meta = NLPModels.NLPModelMeta(
+    n;
+    lvar=NLPModels.get_lvar(nlp),
+    uvar=NLPModels.get_uvar(nlp),
+    x0=NLPModels.get_x0(nlp),
+    y0 = NLPModels.get_y0(nlp) .* scaling_cons,
+    nnzj=NLPModels.get_nnzj(nlp),
+    nnzh=NLPModels.get_nnzh(nlp),
+    ncon=m,
+    lcon=NLPModels.get_lcon(nlp) .* scaling_cons,
+    ucon=NLPModels.get_ucon(nlp) .* scaling_cons,
+    minimize=true,
+  )
+
+  return ScaledModel(
+    nlp,
+    meta,
+    NLPModels.Counters(),
+    scaling_obj,
+    scaling_cons,
+    scaling_jac,
+    buffer_cons,
+  )
+end
+
+function NLPModels.obj(nlp::ScaledModel{T, S}, x::AbstractVector) where {T, S <: AbstractVector{T}}
+  @lencheck nlp.meta.nvar x
+  return nlp.scaling_obj * NLPModels.obj(nlp.nlp, x)
+end
+
+function NLPModels.cons!(nlp::ScaledModel, x::AbstractVector, c::AbstractVector)
```

Comment:
This is more a general comment on future work. We recently split the constraint API to nonlinear and linear. Would it make sense in a future work to have two different scaling for linear and nonlinear constraints?

Reply:
Actually, anyway it would be better to have `cons_lin!` and `cons_nln!` instead of `cons!`

Reply:
I would prefer to keep the interface as is. As far as I understand `cons!` is calling by default `cons_lin!` and `cons_nln!` internally, and here the scaling does not depend on the nature of the constraint.